### ISSUE TITLE
Missing submodules in Nixos

### ISSUE THREAD

**Issue by leenaars** [nixcloud/nixcloud-webservices#64]
It is no longer possible to build nixcloud-webservices with current nixos-unstable:

building Nix...
building the system configuration...
error: attribute 'submodule' missing, at /home/username/repo/nixcloud-webservices/modules/web/core/webserver.nix:66:39
(use '--show-trace' to show detailed location information)

**Comment by leenaars** [comment]
There were some changes in modules related to nixcloud-webservices::

 https://nixos.org/nixos/manual/release-notes.html#sec-release-19.09

**Comment by fgaz** [comment]
This prevents nixcloud-webservices to work on the newly released 20.03

**Comment by Infinisil** [comment]
A couple days ago I had a discussion with @Melkor333 on IRC regarding this: https://logs.nix.samueldr.com/nixos/2020-07-12#1594566523-1594572059; where I also suggested a non-optimal solution

The change in nixpkgs that introduced this problem is my PR https://github.com/NixOS/nixpkgs/pull/75031. I don't think this is a problem with the PR though, because the problematic code in nixcloud-webservices at https://github.com/nixcloud/nixcloud-webservices/blob/3a0767f0536fac811065eb87e6342f27eac085aa/modules/web/core/webserver.nix#L64-L66 (among others) relies on module system internals, which I removed in that PR: https://github.com/NixOS/nixpkgs/pull/75031/files#diff-490491facef0aa6a6dbe8ffc5bd97097L174-L184

After messing with it for a bit, here is how the same functionality can be achieved without relying on internals:
```nix
{
  options.webserver.userOptions = lib.mkOption {
    type = lib.types.submodule options.users.users.type.functor.wrapped.getSubModules;
    default = {};
  };

  config.users.users.webserver = lib.modules.mkAliasAndWrapDefsWithPriority lib.id options.webserver.userOptions;
```

Also, make sure to not use `//` for merging in the module system like in https://github.com/nixcloud/nixcloud-webservices/blob/3a0767f0536fac811065eb87e6342f27eac085aa/modules/web/core/webserver.nix#L120-L122, because it can [lead to problems](https://github.com/NixOS/nixpkgs/pull/72949). Prefer `mkMerge` instead. And to prevent attributes from being changed, a `mkForce` is recommended. So in the end that part looks about like this:
```nix
{
  users.${config.webserver.user} = lib.mkMerge [
    (lib.mkForce { inherit (config.webserver) user group; })
    (lib.modules.mkAliasAndWrapDefsWithPriority lib.id options.webserver.userOptions)
  ];
}
```

To prevent `name` and `group` from being changed you could set `group` and `name` with a `mkForce`.

**Comment by Melkor333** [comment]
Oh wow thank you for taking another look at it!! 😯 I hope I’ll have time next weekend to run the tests with this code and create a PR...

**Comment by fgaz** [comment]
This is still a problem and prevents nixcloud-webservices to work with the newly released 20.09. I'll try to apply the same fix to the remaining places and submit a pr shortly

**Comment by qknight** [comment]
I still have this issue on 19.09 

```
root@mail /e/n/nixpkgs# git bisect bad
5414b4018bf1161ad9bd0e98d0ffbde8aa435fe5 is the first bad commit
commit 5414b4018bf1161ad9bd0e98d0ffbde8aa435fe5
Author: Silvan Mosberger <contact@infinisil.com>
Date:   Thu Dec 5 03:29:51 2019 +0100

    lib/modules: Don't pack submodules specially

    This has the beneficial side effect of allowing paths to be used as modules in
    types.{submodule,submoduleWith}

 lib/default.nix |  2 +-
 lib/modules.nix | 26 ++++++++++----------------
 2 files changed, 11 insertions(+), 17 deletions(-)
```

https://github.com/NixOS/nixpkgs/commit/5414b4018bf1161ad9bd0e98d0ffbde8aa435fe5
